packages/@react-spectrum/tag/src/TagGroup.tsx

```diff
@@ -70,6 +72,78 @@ function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef
   }, state, domRef);
   const {tagGroupProps} = useTagGroup(props);
 
+  let [isCollapsed, setIsCollapsed] = useState(maxRows != null);
+  let [visibleTagCount, setVisibleTagCount] = useValueEffect(gridCollection.size);
+
+  let updateVisibleTagCount = useCallback(() => {
+    let computeVisibleTagCount = () => {
+      // Refs can be null at runtime.
+      let currDomRef: HTMLDivElement | null = domRef.current;
+      if (!currDomRef) {
+        return;
+      }
+
+      let tags = [...currDomRef.children].filter(el => el.tagName !== 'BUTTON') as HTMLDivElement[];
```

Comment:
if you track the width of each tag and also get the width of the button here, you can remove the loop that removes tags one by one, thereby reducing the number of renders from `n+1` to `2`

Reply:
The collapse button doesn't exist yet here, but I've updated it to include a separate function for making space for the collapse button (and passing in the widths there), so number of renders should be 3.